File: microservices/src/IoBuild.Analytics/Interfaces/REST/Assemblers/OwnerDashboardAssembler.cs
using IoBuild.Analytics.Domain.Model.Aggregates;
using IoBuild.Analytics.Interfaces.REST.Resources;

namespace IoBuild.Analytics.Interfaces.REST.Assemblers;

public static class OwnerDashboardAssembler
{
    public static OwnerDashboardResource ToResource(OwnerMetrics metrics)
    {
        return new OwnerDashboardResource(
            metrics.TotalDevices,
            metrics.OnlineDevices,
            metrics.OfflineDevices,
            metrics.AlertsCount,
            metrics.MyUnitsCount,
            metrics.EnergyThisMonth,
            metrics.TemperatureAvg,
            metrics.WaterUsageThisMonth,
            metrics.TemperatureHistory.Select(HistoricalDataPointAssembler.ToResource).ToList(),
            metrics.EnergyHistory.Select(HistoricalDataPointAssembler.ToResource).ToList(),
            metrics.DailyEnergyConsumption.Select(HistoricalDataPointAssembler.ToResource).ToList(),
            metrics.WaterUsageWeekly.Select(HistoricalDataPointAssembler.ToResource).ToList(),
            metrics.DeviceHealthStatus.Select(h => new DeviceHealthStatusResource(h.DeviceId, h.DeviceName, h.Status, h.LastOnline)).ToList(),
            metrics.MyUnitsDetails
        );
    }
}

File: microservices/frontend-docker/IoBuild-Frontend/src/analytics/infrastructure/owner-dashboard.assembler.js
import { OwnerDashboard } from "../domain/model/owner-dashboard.entity.js";
import { HistoricalDataPoint } from "../domain/model/historical-data-point.entity.js";

export class OwnerDashboardAssembler {
    static toEntity(resource) {
        return new OwnerDashboard({
            totalDevices: resource.totalDevices,
            onlineDevices: resource.onlineDevices,
            offlineDevices: resource.offlineDevices,
            alertsCount: resource.alertsCount,
            myUnitsCount: resource.myUnitsCount,
            energyThisMonth: resource.energyThisMonth,
            temperatureAvg: resource.temperatureAvg,
            waterUsageThisMonth: resource.waterUsageThisMonth,
            temperatureHistory: resource.temperatureHistory.map(point => 
                new HistoricalDataPoint({
                    timestamp: new Date(point.timestamp),
                    value: point.value,
                    type: point.type
                })
            ),
            energyHistory: resource.energyHistory.map(point => 
                new HistoricalDataPoint({
                    timestamp: new Date(point.timestamp),
                    value: point.value,
                    type: point.type
                })
            ),
            dailyEnergyConsumption: resource.dailyEnergyConsumption.map(point => 
                new HistoricalDataPoint({
                    timestamp: new Date(point.timestamp),
                    value: point.value,
                    type: point.type
                })
            ),
            waterUsageWeekly: resource.waterUsageWeekly.map(point => 
                new HistoricalDataPoint({
                    timestamp: new Date(point.timestamp),
                    value: point.value,
                    type: point.type
                })
            ),
            deviceHealthStatus: (resource.deviceHealthStatus || []).map(d => ({
                ...d,
                healthPercentage: d.status === 'Online' ? 92 + Math.floor(Math.random() * 8) : 30 + Math.floor(Math.random() * 40),
                type: d.deviceName?.includes('Temperature') ? 'temperature' :
                      d.deviceName?.includes('Energy') ? 'energy' :
                      d.deviceName?.includes('Water') ? 'water' : 'other'
            })),
            myUnitsDetails: resource.myUnitsDetails
        });
    }
}
